### PR TITLE
Improved error handling when GDAL 2.x fails to open existing layer

### PR DESCRIPTION
**Pull request by snorfalorpagus**
This PR provides slightly more informative error messages when failing to open a dataset in read mode. Previous more detailed messages from GDAL could be hidden by a generic (and potentially incorrect) message from Fiona.

I came across this issue in the 2.3.x (svn-trunk) job on Travis, which was raising this error:

```
fiona.errors.FionaValueError: No dataset found at path '/vsimem/c513cc7b5ce848a6a3358d4c0854f9bd.json' using drivers: {'GeoJSON'}
```

This is now replaced by a more correct message:

```
fiona.errors.FionaValueError: `/vsimem/8c013c21ebeb42cabfb0bb93239bf4d4.json' not recognized as a supported file format. using drivers {'GeoJSON'}
```

I think the actual problem is something in GDAL (see upstream issue https://trac.osgeo.org/gdal/ticket/7198 ), but the message Fiona gives is more useful now.

### CODE REVIEW

File: fiona/_shim2.pyx [hunk 1]
```diff
@@ -49,10 +51,12 @@ cdef void* gdal_open_vector(char* path_c, int mode, drivers, options):
 
     open_opts = CSLAddNameValue(open_opts, "VALIDATE_OPEN_OPTIONS", "NO")
 
-    flags = GDAL_OF_VECTOR | mode
     try:
         cogr_ds = GDALOpenEx(
             path_c, flags, <const char *const *>drvs, open_opts, NULL)
```
Comment (sgillies) [inline fiona/_shim2.pyx]: @snorfalorpagus let's wrap the `GDALOpenEx` call in `exc_wrap_pointer()` https://github.com/Toblerity/Fiona/blob/master/fiona/_err.pyx#L222

**Comment by snorfalorpagus** [comment]
This PR has grown a bit...

I've refactored the code used to open files for writing. I also fixed a bug in the condition used to detect GPKG support in the tests. This error was skipping some tests that were actually failing. I don't think writing binary fields has ever worked on Python 2.7, but does not. I think is is ready for review.

**Comment by sgillies** [comment]
I'll review tomorrow morning.

File: fiona/_shim1.pyx [hunk 2]
```diff
@@ -50,7 +50,10 @@ cdef void* gdal_open_vector(const char *path_c, int mode, drivers, options):
                 #collection._driver = name
                 break
     else:
-        cogr_ds = OGROpen(path_c, mode, NULL)
+        try:
+            cogr_ds = exc_wrap_pointer(OGROpen(path_c, mode, NULL))
+        except Exception as exc:
```
Comment (sgillies) [inline fiona/_shim1.pyx]: I think we should catch only `CPLE_BaseError` here so we don't accidentally hide any errors related to malformed paths or bogus modes.

File: fiona/_shim2.pyx [hunk 3]
```diff
@@ -49,11 +53,11 @@ cdef void* gdal_open_vector(char* path_c, int mode, drivers, options):
...
     open_opts = CSLAddNameValue(open_opts, "VALIDATE_OPEN_OPTIONS", "NO")
 
-    flags = GDAL_OF_VECTOR | mode
     try:
-        cogr_ds = GDALOpenEx(
-            path_c, flags, <const char *const *>drvs, open_opts, NULL)
-        return cogr_ds
+        return exc_wrap_pointer(GDALOpenEx(
+            path_c, flags, <const char *const *>drvs, open_opts, NULL))
+    except Exception as exc:
```
Comment (sgillies) [inline fiona/_shim2.pyx]: `CPLE_BaseError` here too.

File: fiona/ogrext.pyx [hunk 4]
```diff
@@ -335,6 +335,10 @@ cdef class OGRFeatureBuilder:
                 hh, mm, ss = value.hour, value.minute, value.second
                 OGR_F_SetFieldDateTime(
                     cogr_feature, i, 0, 0, 0, hh, mm, ss, 0)
+            elif isinstance(value, bytes) and schema_type == "bytes":
+                string_c = value
```
Comment (sgillies) [inline fiona/ogrext.pyx]: @snorfalorpagus I think this is a good change, if we try to set a not-binary field to bytes we'll get a more opaque OGR error.

File: fiona/ogrext.pyx [hunk 5]
```diff
@@ -420,26 +420,30 @@ cdef class Session:
...
-                drivers = set([collection._driver])
-            elif collection.enabled_drivers:
-                drivers = set(collection.enabled_drivers)
-            else:
-                drivers = None
-
-            self.cogr_ds = gdal_open_vector(path_c, 0, drivers, kwargs)
+        message = None
+        try:
+            with cpl_errs:
```
Comment (sgillies) [inline fiona/ogrext.pyx]: Could we use `exc_wrap_pointer` instead of `cpl_errs`?

File: fiona/ogrext.pyx [hunk 6]
```diff
@@ -420,26 +420,30 @@ cdef class Session:
...
 
         if self.cogr_ds == NULL:
+            if not message:
+                message = "No dataset found at path '{}'".format(collection.path)
+
             raise FionaValueError(
-                "No dataset found at path '%s' using drivers: %s" % (
-                    collection.path,
-                    drivers or '*'))
+                message + " using drivers {}".format(drivers or "*"))
```
Comment (sgillies) [inline fiona/ogrext.pyx]: 👍

**Comment by snorfalorpagus** [comment]
I tried to apply these changes to `_listlayers` but am getting some odd behavior:

```
>>> _listlayers("abc.json")
Traceback (most recent call last):
  File "fiona\_shim.pyx", line 63, in fiona._shim.gdal_open_vector
    return exc_wrap_pointer(GDALOpenEx(
  File "fiona\_err.pyx", line 229, in fiona._err.exc_wrap_pointer
    raise exc
fiona._err.CPLE_OpenFailedError: abc.json: No such file or directory

During handling of the above exception, another exception occurred:

fiona.errors.DriverIOError: abc.json: No such file or directory
Exception ignored in: 'fiona._shim.gdal_open_vector'
fiona.errors.DriverIOError: abc.json: No such file or directory
```

Can you hold off on this until I've worked this issue out.

**Comment by snorfalorpagus** [comment]
Every time I think I've fixed something, something else breaks!

**Comment by sgillies** [comment]
Don't sweat it @snorfalorpagus ! No rush.

**Comment by snorfalorpagus** [comment]
I think I'm going to split some of these commits out into smaller PRs to make it more manageable.

**Pull request closed without merging**